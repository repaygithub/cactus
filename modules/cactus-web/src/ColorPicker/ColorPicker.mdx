---
name: ColorPicker
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import ColorPicker from './ColorPicker'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'
import { livePreviewStyle } from '../helpers/constants'

# ColorPicker

The `ColorPicker` is a form element designed to allow a user to select a color with hex, rgb, or hue and saturation sliders.

### Try It Out

export const code = `
  <ColorPicker id="my-picker" name="color" />
`

<LiveProvider code={code} scope={{ ColorPicker }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Best practices

- Pass a meaningful name and ID
- Use the `phrases` prop to set text & labels. This object should contain any of the following keys, although all are optional.
  - `hexLabel`: Defaults to "Hex".
  - `redLabel`: Defaults to "R".
  - `greenLabel`: Defaults to "G".
  - `blueLabel`: Defaults to "B".
  - `triggerLabel`: Label for the palette icon button. Defaults to "Click to open the color picker".
  - `applyLabel`: Defaults to "Apply".
  - `cancelLabel`: Defaults to "Cancel".

## Basic usage

- If the hex input rendered outside the popup is changed, `onChange` will be called as soon as a valid hex string is set.
- If the color is changed using any of the elements in the popup, `onChange` will only be called when the apply button is clicked.
- If the cancel button is clicked, the color picker's state will revert to whatever value is passed through the props.
- The `ColorPicker` value can be controlled with any of the following types of values:
  - Hex string - ex: `00FF00` or `#00FF00`
  - HSL object - Hue should stay between 0 and 360 while saturation and lightness should stay between 0 and 1. ex: `{ h: 120, s: 1, l: 0.5 }`
  - HSV object - Hue should stay between 0 and 360 while saturation and value should stay between 0 and 1. ex: `{ h: 120, s: 1, v: 1 }`
  - RGB object - All values should be between 0 and 255. ex: `{r: 0, g: 255, b: 0}`

## Properties

<PropsTable of={ColorPicker} />
