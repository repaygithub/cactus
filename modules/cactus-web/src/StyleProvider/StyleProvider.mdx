---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
