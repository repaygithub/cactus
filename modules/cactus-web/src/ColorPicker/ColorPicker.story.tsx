import { select } from '@storybook/addon-knobs'
import { Meta } from '@storybook/react/types-6-0'
import React, { useState } from 'react'

import ColorPicker from './ColorPicker'

export default {
  title: 'ColorPicker',
  component: ColorPicker,
} as Meta

const storyParams = {
  cactus: { overrides: { alignItems: 'start', paddingTop: '32px' } },
}

const eventLoggers = {
  onChange: (e: any) => {
    console.log(`onChange '${e.target.name}':`)
    console.log(e.target.value)
  },
  onFocus: (e: any) => console.log('onFocus:', e.target.name),
  onBlur: (e: any) => console.log('onBlur:', e.target.name),
}

export const BasicUsage = (): React.ReactElement => (
  <ColorPicker id="color-picker" name="color" {...eventLoggers} />
)

BasicUsage.parameters = storyParams

export const Controlled = (): React.ReactElement => {
  const [state, setState] = useState({
    hsl: { h: 120, s: 1, l: 0.5 },
    hsv: { h: 120, s: 1, v: 1 },
    rgb: { r: 0, g: 255, b: 0 },
    hex: '00FF00',
  })
  const controlledBy = select('controlled by', ['hsl', 'hsv', 'rgb', 'hex'], 'hsl')
  return (
    <ColorPicker
      id="color-picker"
      name="color"
      value={state[controlledBy]}
      {...eventLoggers}
      onChange={(event) => {
        if (event.currentTarget.value) {
          setState(event.currentTarget.value)
          console.log(`onChange: ${event.currentTarget.name}:`)
          console.log(event.currentTarget.value)
        }
      }}
    />
  )
}

Controlled.parameters = storyParams

export const Disabled = (): React.ReactElement => (
  <ColorPicker id="color-picker" name="disabled" disabled />
)
