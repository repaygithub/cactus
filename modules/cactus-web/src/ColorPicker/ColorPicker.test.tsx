import { act, fireEvent, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'

import animationRender from '../../tests/helpers/animationRender'
import { StyleProvider } from '../StyleProvider/StyleProvider'
import ColorPicker from './ColorPicker'

describe('component: ColorPicker', () => {
  describe('can be controlled', () => {
    test('with hsl', () => {
      const { getAllByLabelText, getByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="color-test" value={{ h: 120, s: 1, l: 0.5 }} />
        </StyleProvider>
      )

      expect((getAllByLabelText('Hex')[0] as HTMLInputElement).value).toBe('#00FF00')
      expect((getByLabelText('R') as HTMLInputElement).value).toBe('0')
      expect((getByLabelText('G') as HTMLInputElement).value).toBe('255')
      expect((getByLabelText('B') as HTMLInputElement).value).toBe('0')
    })

    test('with hsv', () => {
      const { getAllByLabelText, getByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="what-color" value={{ h: 120, s: 1, v: 1 }} />
        </StyleProvider>
      )

      expect((getAllByLabelText('Hex')[0] as HTMLInputElement).value).toBe('#00FF00')
      expect((getByLabelText('R') as HTMLInputElement).value).toBe('0')
      expect((getByLabelText('G') as HTMLInputElement).value).toBe('255')
      expect((getByLabelText('B') as HTMLInputElement).value).toBe('0')
    })

    test('with rgb', () => {
      const { getAllByLabelText, getByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="color-me" value={{ r: 0, g: 255, b: 0 }} />
        </StyleProvider>
      )

      expect((getAllByLabelText('Hex')[0] as HTMLInputElement).value).toBe('#00FF00')
      expect((getByLabelText('R') as HTMLInputElement).value).toBe('0')
      expect((getByLabelText('G') as HTMLInputElement).value).toBe('255')
      expect((getByLabelText('B') as HTMLInputElement).value).toBe('0')
    })

    test('with hex', () => {
      const { getAllByLabelText, getByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="almost-there" value="00FF00" />
        </StyleProvider>
      )

      expect((getAllByLabelText('Hex')[0] as HTMLInputElement).value).toBe('#00FF00')
      expect((getByLabelText('R') as HTMLInputElement).value).toBe('0')
      expect((getByLabelText('G') as HTMLInputElement).value).toBe('255')
      expect((getByLabelText('B') as HTMLInputElement).value).toBe('0')
    })
  })

  test('can open portal', async () => {
    const { getByLabelText, getAllByLabelText, getByRole } = render(
      <StyleProvider>
        <ColorPicker id="color-picker" name="pick-a-color" />
      </StyleProvider>
    )

    const trigger = getByLabelText('Click to open the color picker')
    userEvent.click(trigger)
    expect(getByRole('dialog')).toBeInTheDocument()
    // Should focus the hex input when initially opened
    await animationRender()
    expect(document.activeElement).toBe(getAllByLabelText('Hex')[1])
  })

  describe('calls onChange', () => {
    test('when apply button is clicked', () => {
      const onChange = jest.fn()
      const { getByLabelText, getByText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="rainbow" onChange={onChange} />
        </StyleProvider>
      )

      const trigger = getByLabelText('Click to open the color picker')
      userEvent.click(trigger)

      const redInput = getByLabelText('R') as HTMLInputElement
      const greenInput = getByLabelText('G') as HTMLInputElement
      const blueInput = getByLabelText('B') as HTMLInputElement

      userEvent.clear(redInput)
      userEvent.type(redInput, '0')
      fireEvent.blur(redInput)

      userEvent.clear(greenInput)
      userEvent.type(greenInput, '0')
      fireEvent.blur(greenInput)

      userEvent.clear(blueInput)
      userEvent.type(blueInput, '255')
      fireEvent.blur(blueInput)

      expect(onChange).not.toHaveBeenCalled()

      const applyButton = getByText('Apply')

      userEvent.click(applyButton)

      expect(onChange).toHaveBeenCalledTimes(1)
    })

    test('when outer hex input is changed', () => {
      const onChange = jest.fn()
      const { getAllByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="rainbow" onChange={onChange} />
        </StyleProvider>
      )

      const outerHexInput = getAllByLabelText('Hex')[0]
      userEvent.clear(outerHexInput)
      userEvent.type(outerHexInput, 'FF0000')

      expect(onChange).toHaveBeenCalledTimes(1)
    })
  })

  describe('does not call onChange', () => {
    test('when cancel button is clicked', () => {
      const onChange = jest.fn()
      const { getByLabelText, getByText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="rainbow" onChange={onChange} />
        </StyleProvider>
      )

      const trigger = getByLabelText('Click to open the color picker')
      userEvent.click(trigger)

      const redInput = getByLabelText('R') as HTMLInputElement
      const greenInput = getByLabelText('G') as HTMLInputElement
      const blueInput = getByLabelText('B') as HTMLInputElement

      userEvent.clear(redInput)
      userEvent.type(redInput, '0')
      fireEvent.blur(redInput)

      userEvent.clear(greenInput)
      userEvent.type(greenInput, '0')
      fireEvent.blur(greenInput)

      userEvent.clear(blueInput)
      userEvent.type(blueInput, '255')
      fireEvent.blur(blueInput)

      const cancelButton = getByText('Cancel')

      userEvent.click(cancelButton)

      expect(onChange).not.toHaveBeenCalled()
    })
  })

  describe('calls onFocus', () => {
    test('when outer hex input is focused', () => {
      const onFocus = jest.fn()
      const { getAllByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="rainbow" onFocus={onFocus} />
        </StyleProvider>
      )

      const outerHexInput = getAllByLabelText('Hex')[0]
      fireEvent.focus(outerHexInput)
      expect(onFocus).toHaveBeenCalledTimes(1)
    })

    test('when icon button is clicked', () => {
      const onFocus = jest.fn()
      const onBlur = jest.fn()
      const { getByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="rainbow" onFocus={onFocus} onBlur={onBlur} />
        </StyleProvider>
      )

      const trigger = getByLabelText('Click to open the color picker')
      userEvent.click(trigger)
      expect(onFocus).toHaveBeenCalledTimes(1)
      expect(onBlur).toHaveBeenCalledTimes(0)
    })
  })

  describe('calls onBlur', () => {
    test('when the outer hex input is blurred', async () => {
      const onBlur = jest.fn()
      const { getAllByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="rainbow" onBlur={onBlur} />
        </StyleProvider>
      )

      const outerHexInput = getAllByLabelText('Hex')[0]
      fireEvent.focus(outerHexInput)
      fireEvent.blur(outerHexInput)
      await animationRender()
      expect(onBlur).toHaveBeenCalledTimes(1)
    })

    test('when the cancel button is clicked', async () => {
      const onBlur = jest.fn()
      const { getByLabelText, getByText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="rainbow" onBlur={onBlur} />
        </StyleProvider>
      )

      const trigger = getByLabelText('Click to open the color picker')
      userEvent.click(trigger)

      const cancelButton = getByText('Cancel')
      userEvent.click(cancelButton)
      await animationRender()
      expect(onBlur).toHaveBeenCalledTimes(1)
    })

    test('when the apply button is clicked', async () => {
      const onBlur = jest.fn()
      const { getByLabelText, getByText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="rainbow" onBlur={onBlur} />
        </StyleProvider>
      )

      const trigger = getByLabelText('Click to open the color picker')
      userEvent.click(trigger)

      const applyButton = getByText('Apply')
      userEvent.click(applyButton)
      await animationRender()
      expect(onBlur).toHaveBeenCalledTimes(1)
    })
  })

  describe('can set color', () => {
    test('using the outer hex input', () => {
      const { getAllByLabelText, getByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="does-anyone-read-this" />
        </StyleProvider>
      )

      const outerHexInput = getAllByLabelText('Hex')[0] as HTMLInputElement
      const innerHexInput = getAllByLabelText('Hex')[1] as HTMLInputElement
      userEvent.clear(outerHexInput)
      userEvent.type(outerHexInput, 'FF0000')
      fireEvent.blur(outerHexInput)
      expect(innerHexInput.value).toBe('#FF0000')
      expect((getByLabelText('R') as HTMLInputElement).value).toBe('255')
      expect((getByLabelText('G') as HTMLInputElement).value).toBe('0')
      expect((getByLabelText('B') as HTMLInputElement).value).toBe('0')
    })

    test('using the inner hex input', () => {
      const { getAllByLabelText, getByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="cool-color" />
        </StyleProvider>
      )

      const trigger = getByLabelText('Click to open the color picker')
      userEvent.click(trigger)

      const outerHexInput = getAllByLabelText('Hex')[0] as HTMLInputElement
      const innerHexInput = getAllByLabelText('Hex')[1] as HTMLInputElement
      userEvent.clear(innerHexInput)
      userEvent.type(innerHexInput, 'FF0000')
      fireEvent.blur(innerHexInput)
      expect(outerHexInput.value).toBe('#FF0000')
      expect((getByLabelText('R') as HTMLInputElement).value).toBe('255')
      expect((getByLabelText('G') as HTMLInputElement).value).toBe('0')
      expect((getByLabelText('B') as HTMLInputElement).value).toBe('0')
    })

    test('using the rgb inputs', () => {
      const { getAllByLabelText, getByLabelText } = render(
        <StyleProvider>
          <ColorPicker id="color-picker" name="coooooolor" />
        </StyleProvider>
      )

      const trigger = getByLabelText('Click to open the color picker')
      userEvent.click(trigger)

      const outerHexInput = getAllByLabelText('Hex')[0] as HTMLInputElement
      const innerHexInput = getAllByLabelText('Hex')[1] as HTMLInputElement
      const redInput = getByLabelText('R') as HTMLInputElement
      const greenInput = getByLabelText('G') as HTMLInputElement
      const blueInput = getByLabelText('B') as HTMLInputElement

      userEvent.clear(redInput)
      userEvent.type(redInput, '0')
      fireEvent.blur(redInput)

      userEvent.clear(greenInput)
      userEvent.type(greenInput, '0')
      fireEvent.blur(greenInput)

      userEvent.clear(blueInput)
      userEvent.type(blueInput, '255')
      fireEvent.blur(blueInput)

      expect(outerHexInput.value).toBe('#0000FF')
      expect(innerHexInput.value).toBe('#0000FF')
      expect(redInput.value).toBe('0')
      expect(greenInput.value).toBe('0')
      expect(blueInput.value).toBe('255')
    })
  })

  test('snapshot', () => {
    const { container } = render(
      <StyleProvider>
        <ColorPicker id="color-picker" name="color" />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })
})
