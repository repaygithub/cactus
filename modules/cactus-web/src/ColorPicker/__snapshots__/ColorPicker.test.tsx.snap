// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: ColorPicker snapshot 1`] = `
.c15 {
  position: relative;
  padding: 2px 30px;
  outline: none;
  cursor: pointer;
  overflow: visible;
  box-sizing: border-box;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 18px;
  line-height: 1.5;
  border: 1px solid;
  border-radius: 8px;
  background-color: hsl(200,96%,35%);
  color: hsl(0,0%,100%);
  border-color: hsl(200,96%,35%);
}

.c15::-moz-focus-inner {
  border: 0;
}

.c15:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 1px solid;
  border-radius: 8px;
  border-color: hsl(200,96%,35%);
  box-sizing: border-box;
}

.c15 svg {
  display: inline-block;
  margin-top: -4px;
}

.c15 .sc-hkePoP {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c15:hover {
  background-color: hsl(200,96%,11%);
  color: hsl(0,0%,100%);
  border-color: hsl(200,96%,11%);
}

.c3 {
  box-sizing: border-box;
  width: 100%;
}

.c9 {
  box-sizing: border-box;
  width: 100%;
  margin-top: 40px;
}

.c12 {
  box-sizing: border-box;
  width: 100%;
  margin-top: 16px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 1px;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
  position: relative;
  overflow: visible;
  color: hsl(200,96%,11%);
  font-size: 16px;
}

.c1::-moz-focus-inner {
  border: 0;
}

.c1:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 8px);
  width: calc(100% + 8px);
  border: 1px solid;
  border-radius: 8px;
  border-color: hsl(200,96%,35%);
  box-sizing: border-box;
}

.c1:hover {
  color: hsl(200,96%,35%);
}

.c14 {
  font-size: 18px;
  line-height: 1.5;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(353,84%,44%);
}

.c14:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c14:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 1px solid hsl(200,96%,35%);
  border-radius: 8px;
}

.c14::-moz-focus-inner {
  border: 0;
}

.c14 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

.c2 {
  display: none;
  box-sizing: border-box;
  position: fixed;
  z-index: 1000;
  width: 320px;
  padding: 16px;
  background-color: hsl(0,0%,100%);
  border-radius: 0 20px 10px 10px;
  box-shadow: 0px 3px 8px hsla(200,96%,35%,0.3);
  overflow: hidden;
  outline: none;
}

.c2 input {
  border: 1px solid hsl(200,10%,20%);
  border-radius: 8px;
  height: 32px;
  outline: none;
  padding: 0px 15px 0px 15px;
}

.c2 input:focus {
  border: 1px solid hsl(200,96%,35%);
}

.c2 label {
  position: absolute;
  top: -26px;
  left: 16px;
}

.c2 .hex-wrapper > div > input {
  width: 5em;
}

.c2 .rgb-wrapper > div > input {
  width: 2.5em;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid hsl(200,10%,20%);
  border-radius: 8px;
  height: 36px;
  outline: none;
  padding: 0 16px 0 12px;
  min-width: 160px;
}

.c0:focus-within {
  border-color: hsl(200,96%,35%);
}

.c0 input {
  border: none;
  outline: none;
  padding: 0;
  background-color: transparent;
  font-size: inherit;
  width: 5em;
}

.c5 {
  position: relative;
  width: 288px;
  height: 200px;
}

.c7 {
  position: relative;
  width: 288px;
  height: 16px;
  margin-top: 32px;
}

.c6 {
  box-sizing: border-box;
  width: 32px;
  height: 32px;
  border-radius: 50%;
  background-color: transparent;
  border: 8px solid hsl(0,0%,100%);
  box-shadow: 0 0 3px hsl(0,0%,70%);
  -webkit-transform: translate(-24px,-8px);
  -ms-transform: translate(-24px,-8px);
  transform: translate(-24px,-8px);
}

.c8 {
  box-sizing: border-box;
  width: 32px;
  height: 32px;
  border-radius: 50%;
  background-color: transparent;
  border: 8px solid hsl(0,0%,100%);
  box-shadow: 0 0 3px hsl(0,0%,70%);
  -webkit-transform: translate(-8px,-8px);
  -ms-transform: translate(-8px,-8px);
  transform: translate(-8px,-8px);
}

@media screen and (min-width:1024px) {
  .c15 {
    font-size: 18px;
    line-height: 1.5;
  }
}

@media screen and (min-width:1024px) {
  .c14 {
    font-size: 18px;
    line-height: 1.5;
  }
}

<div>
  <div
    class="c0"
    id="color-picker"
  >
    <button
      aria-controls="color-picker-popup"
      aria-haspopup="dialog"
      aria-label="Click to open the color picker"
      class="c1"
      id="color-picker-button"
      role="button"
      type="button"
      variant="standard"
    >
      <svg
        fill="currentcolor"
        height="14"
        viewBox="0 0 26 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.0993 5.94055C16.1785 2.01982 9.8215 2.01982 5.90076 5.94055C1.98002 9.86129 1.98002 16.2183 5.90076 20.1391C6.25918 20.4975 6.84546 20.4975 7.20389 20.1391C7.56231 19.7806 7.56231 19.1944 7.20389 18.8359C6.43889 18.0709 6.43889 16.8328 7.20389 16.0679C7.96889 15.3029 9.20697 15.3029 9.97197 16.0679L14.0432 20.1391C15.7171 21.813 18.4254 21.813 20.0993 20.1391C24.02 16.2183 24.02 9.86129 20.0993 5.94055ZM4.48655 4.52634C9.18833 -0.175447 16.8117 -0.175447 21.5135 4.52634C26.2153 9.22813 26.2153 16.8515 21.5135 21.5533C19.0585 24.0082 15.0839 24.0082 12.629 21.5533L9.05463 17.9789C9.73184 19.1019 9.58633 20.585 8.6181 21.5533C7.47863 22.6927 5.62602 22.6927 4.48655 21.5533C-0.215242 16.8515 -0.215242 9.22813 4.48655 4.52634ZM17.376 7.97009C17.2143 7.80557 16.94 7.7968 16.7606 7.97616L16.7546 7.9822L16.7546 7.98218C16.5867 8.14723 16.5867 8.42648 16.7546 8.59154L16.7667 8.60362C16.9317 8.77152 17.211 8.77152 17.376 8.60362L17.3881 8.59154C17.556 8.42648 17.556 8.14723 17.3881 7.98218L17.376 7.97009ZM18.7963 6.56194C17.8447 5.60057 16.3014 5.61026 15.3494 6.55894C14.3894 7.506 14.3884 9.06343 15.3464 10.0118C16.2938 10.9688 17.8489 10.9688 18.7963 10.0118C19.7533 9.06443 19.7533 7.5093 18.7963 6.56194ZM19.7612 13.0398C19.7612 13.0618 19.7434 13.0796 19.7214 13.0796C19.6994 13.0796 19.6816 13.0618 19.6816 13.0398C19.6816 13.0178 19.6994 13 19.7214 13C19.7434 13 19.7612 13.0178 19.7612 13.0398ZM19.7214 11.0796C18.6388 11.0796 17.7612 11.9572 17.7612 13.0398C17.7612 14.1224 18.6388 15 19.7214 15C20.804 15 21.6816 14.1224 21.6816 13.0398C21.6816 11.9572 20.804 11.0796 19.7214 11.0796ZM17.7276 17.1364C17.3692 16.778 16.7829 16.778 16.4245 17.1364C16.0661 17.4948 16.0661 18.0811 16.4245 18.4395C16.7829 18.798 17.3692 18.798 17.7276 18.4395C18.0861 18.0811 18.0861 17.4948 17.7276 17.1364ZM15.0103 15.7222C16.1498 14.5827 18.0024 14.5827 19.1419 15.7222C20.2813 16.8617 20.2813 18.7143 19.1419 19.8537C18.0024 20.9932 16.1498 20.9932 15.0103 19.8537C13.8708 18.7143 13.8708 16.8617 15.0103 15.7222ZM10.9899 6.95836C11.1692 6.779 11.4435 6.78777 11.6052 6.95229L11.6052 6.95231L11.6113 6.95836C11.7906 7.13772 11.7819 7.412 11.6173 7.57374L11.6173 7.57371L11.6113 7.57975C11.4319 7.75911 11.1576 7.75035 10.9959 7.58582L10.9838 7.57374C10.8193 7.412 10.8105 7.13772 10.9899 6.95836ZM9.57566 5.54414C10.5283 4.59148 12.0762 4.58173 13.0285 5.54718C13.9782 6.50009 13.9869 8.04564 13.0225 8.99697C12.0706 9.94565 10.5272 9.95534 9.57564 8.99398C8.61327 8.04139 8.62399 6.49581 9.57566 5.54414Z"
        />
      </svg>
    </button>
    <div
      style="position: relative;"
    >
      <input
        id="rc-editable-input-81"
        spellcheck="false"
        value="#FF0000"
      />
      <label
        for="rc-editable-input-81"
        style="visibility: hidden; width: 0px; height: 0px; position: absolute;"
      >
        Hex
      </label>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="color-picker-button"
    class="c2"
    id="color-picker-popup"
    role="dialog"
    tabindex="-1"
  >
    <div
      data-cactus-focus-guard="start"
      tabindex="0"
    />
    <div
      class="c3 c4"
      width="100%"
    >
      <div
        class="c5"
      >
        <div
          style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px; background: rgb(255, 0, 0); border-radius: 8px;"
        >
          <style>
            
          .saturation-white {
            background: -webkit-linear-gradient(to right, #fff, rgba(255,255,255,0));
            background: linear-gradient(to right, #fff, rgba(255,255,255,0));
          }
          .saturation-black {
            background: -webkit-linear-gradient(to top, #000, rgba(0,0,0,0));
            background: linear-gradient(to top, #000, rgba(0,0,0,0));
          }
        
          </style>
          <div
            class="saturation-white"
            style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px; border-radius: 8px;"
          >
            <div
              class="saturation-black"
              style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px; border-radius: 8px;"
            />
            <div
              style="position: absolute; top: 0%; left: 100%; cursor: default;"
            >
              <div
                class="c6"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="c7"
      >
        <div
          style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px; border-radius: 8px;"
        >
          <div
            class="hue-horizontal"
            style="padding: 0px 2px; position: relative; height: 100%; border-radius: 8px;"
          >
            <style>
              
            .hue-horizontal {
              background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
                33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to right, #f00 0%, #ff0
                17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }

            .hue-vertical {
              background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
                #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
                #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }
          
            </style>
            <div
              style="position: absolute; left: 0%;"
            >
              <div
                class="c8"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="c9 c10 hex-wrapper"
        width="100%"
      >
        <div
          style="position: relative;"
        >
          <input
            id="rc-editable-input-82"
            spellcheck="false"
            value="#FF0000"
          />
          <label
            for="rc-editable-input-82"
          >
            Hex
          </label>
        </div>
        <div
          class="c9 c11 rgb-wrapper"
          width="100%"
        >
          <div
            style="position: relative;"
          >
            <input
              id="rc-editable-input-83"
              spellcheck="false"
              value="255"
            />
            <label
              for="rc-editable-input-83"
            >
              R
            </label>
          </div>
          <div
            style="position: relative;"
          >
            <input
              id="rc-editable-input-84"
              spellcheck="false"
              value="0"
            />
            <label
              for="rc-editable-input-84"
            >
              G
            </label>
          </div>
          <div
            style="position: relative;"
          >
            <input
              id="rc-editable-input-85"
              spellcheck="false"
              value="0"
            />
            <label
              for="rc-editable-input-85"
            >
              B
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c12 c13"
      width="100%"
    >
      <button
        class="c14"
        type="button"
      >
        Cancel
      </button>
      <button
        aria-live="assertive"
        class="c15"
        type="button"
      >
        <span>
          Apply
        </span>
      </button>
    </div>
    <div
      data-cactus-focus-guard="end"
      tabindex="0"
    />
  </div>
</div>
`;
